test/test_client.rb

```diff
@@ -53,31 +56,30 @@ def test_tries_increments_with_each_call
   end
 
   def test_metadata_when_not_initialized_with_metadata
-    new_raw_metadata = stub(:new_raw_metadata)
-
-    client = Rets::Client.new(:login_url => "http://example.com")
-    client.stubs(:retrieve_metadata).returns(new_raw_metadata)
+    client = Rets::Client.new(
+      :capabilities => { "GetMetadata" => "http://example.com/metadata" },
+      :http_client  => MockHttpClient.new("http://example.com/metadata" => "METADATA_CONTENT")
+    )
 
-    assert_same new_raw_metadata, client.metadata.marshal_dump
+    assert_equal client.metadata(["SYSTEM"]).marshal_dump, {"SYSTEM"=>"METADATA_CONTENT"}
   end
 
   def test_initialize_with_old_metadata_cached_contstructs_new_metadata_from_request
-    metadata = stub(:current? => false)
-    new_raw_metadata = stub(:new_raw_metadata)
-
-    client = Rets::Client.new(:login_url => "http://example.com", :metadata => metadata)
-    client.stubs(:capabilities).returns({})
-    client.stubs(:retrieve_metadata).returns(new_raw_metadata)
+    client = Rets::Client.new(
+      :capabilities => { "GetMetadata" => "http://example.com/metadata" },
+      :http_client  => MockHttpClient.new("http://example.com/metadata" => "NEW_METADATA_CONTENT"),
+      :metadata     => stub(:current? => false)
+    )
```

Comment:
I feel like there has to be a better way of making the code self-explanatory.

When calling `client.metadata` we check if the cached metadata is up to date (hence the `stub(:current? => false)`.

If the cached metadata is out of date then we try to refetch it. To fetch anything we have to know the url, so we have to have cached `capabilities`. Using the capabilities we can make a http request for the metadata.